parser.py

```diff
@@ -86,31 +97,57 @@ def __rpow__(self, power):
 
 
 class UnaryOp(ArithmeticOp):
-    def __str__(self):
+    def __init__(self, func, exp):
+        super().__init__(func, exp)
+
+    def __repr__(self):
         return "UnaryOp(func={}, exp={})".format(*self.tuple)
 
 
 class BinaryOp(ArithmeticOp):
-    def __str__(self):
+    def __init__(self, func, exp1, exp2):
+        super().__init__(func, exp1, exp2)
+
+    def __repr__(self):
         return "BinaryOp(func={}, exp1={}, exp2={})".format(*self.tuple)
 
 
 class Op:
-    def __init__(self, name, params, wires, string):
+    def __init__(self, name, params, wires):
         self.name = name
         self.params = params
         self.wires = wires
-        self.string = string
 
-    def __str__(self):
-        return self.string
+    def __repr__(self):
+        if self.params:
+            return f"{self.name}({','.join(str(p) for p in self.params)}) {','.join(str(w) for w in self.wires)};"
+
+        return f"{self.name} {','.join(str(w) for w in self.wires)};"
```

Comment:
All serialization logic is now contained within each atomic object. Later, we should probably have separate `.serialize()` and `__repr__()` methods.